compiler/rustc_next_trait_solver/src/solve/effect_goals.rs

```diff
@@ -54,21 +57,34 @@ where
                     return Err(NoSolution);
                 }
 
-                ecx.probe_trait_candidate(source).enter(|ecx| {
+                return ecx.probe_trait_candidate(source).enter(|ecx| {
                     let assumption_trait_pred = ecx.instantiate_binder_with_infer(host_clause);
                     ecx.eq(
                         goal.param_env,
                         goal.predicate.trait_ref,
                         assumption_trait_pred.trait_ref,
                     )?;
                     then(ecx)
-                })
-            } else {
-                Err(NoSolution)
+                });
+            }
+
+            // PERF(sized-hierarchy): Sizedness supertraits aren't elaborated to improve perf, so
+            // check for a `Sized` subtrait when looking for `MetaSized`. `PointeeSized` bounds
+            // are syntactic sugar for a lack of bounds so don't need this.
+            if ecx.cx().is_lang_item(goal_did, TraitSolverLangItem::MetaSized)
+                && ecx.cx().is_lang_item(host_clause_did, TraitSolverLangItem::Sized)
```

Comment:
this does not check that the clause and the goal have the same polarity cc https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/perf.20of.20sized.20hierarchy.20changes.20.28.23137944.29/near/512855088

Reply:
Fixed